design/handle-backup-of-volumes-by-resources-filters.md

```diff
@@ -0,0 +1,223 @@
+ 
+The storage part defined rules including `pv` and `volume`, which correspond to `Kopia, Restic` and `Volume snapshot`.
+ 
+A filter in storage with a specific key and empty value, which means the value matches any value. For example, if the `storage.pv.exclude.persistentVolumeSource.nfs` is `{}` it means if `NFS` is used as `persistentVolumeSource` in Persistent Volume will be skipped no matter what the NFS server or NFS Path is,  
+ 
+A filter may have multiple values, all the values are concatenated by commas. For example, the `storage.pv.include.storageClassName` is `gp2, ebs-sc` which means Persistent Volume with gp2 or ebs-sc storage class both will be back up.
+ 
+The size of each single filter value should limit to 256 bytes in case of an unfriendly long variable assignment.
+ 
+If user defined pv filter rules but used Kopia or Restic to do a backup, the backup will fail in validating the resource filter configuration. Same as the situation if using defined volume filter rules but using CSI or plugins to take volume snapshots.
+ 
+For capacity in `pv` or size in `volume`, the value should include the lower value and upper value concatenated by commas. And it has several combinations below:
+- "0,5Gi" or "0Gi,5Gi" which means capacity or size matches from 0 to 5Gi, including value 0 and value 5Gi
+- ",5Gi" which is equal to "0,5Gi"
+- "5Gi," which means capacity or size matches larger than 5Gi, including value 5Gi
+- "5Gi" which is not supported and will be failed in validating configuration.
+ 
+### Filter Reference
+Currently, resources filters are defined in `BackupSpec` struct, it will be more and more bloated with adding more and more filters which makes the size of `Backup` CR bigger and bigger, so we want to store the resources rules in configmap, and `Backup` CRD reference to current configmap.
+
+the `configmap` would be like this:
+```
+apiVersion: v1
+data:
+ filter.json: |-
+   {
+       "resources": {
+           "include": {
+              ...
+           },
+           "exclude": {
+              ...
+           }
+       },
+       "storage": {
+           "pv": {
+               "include": {
+                  ...
+               },
+               "exclude": {
+                  ...
+               }
+           },
+           "volume": {
+               "include": {
+                  ...
+               },
+               "exclude": {
+                  ...
+               }
+           }
+       }
+   }
+kind: ConfigMap
+metadata:
+ creationTimestamp: "2023-01-16T14:08:12Z"
+ name: cm-backup01
+ namespace: default
+ resourceVersion: "17891025"
+ uid: b73e7f76-fc9e-4e72-8e2e-79db717fe9f1
+```
+
+a new variable `filterConfigmap` would be added into `BackupSpec`, it's value is assigned with current resources filters configmap
+```
+apiVersion: velero.io/v1
+kind: Backup
+metadata:
+  name: backup-1
+  spec:
+    filterConfigmap: cm-backup01
```

Comment:
To make it a little more flexible, how about 
```
XXXfilter:
  ref-type: Configmap
  ref: cm-backup01
```
As for what the `XXX` should be, I'm not very sure it should be `resource` or `object` or some other term.  The term `resource` seems ambiguous in k8s, sometimes it means a _kind_ of resource, for example, `pod` or `storageclass` are referred to as resource names, but `pod-abc` is the name of an instance of `pod` resource, and the goal for this design is to introduce a filter to `persistentvolume`, which is more detailed than resource level.

Reply:
If we want to extend the filter policy, Maybe `resourcesFilter` is good?  or if only focus on `persistentvolume` maybe it could be `PVfilter` or `volumeFilter`?